crates/sui-rpc-loadgen/src/payload/get_checkpoints.rs

```diff
@@ -0,0 +1,141 @@
+        let clients = self.get_clients().await?;
+
+        let end_checkpoints: Vec<CheckpointSequenceNumber> =
+            join_all(clients.iter().map(|client| async {
+                match op.end {
+                    Some(e) => e,
+                    None => client
+                        .read_api()
+                        .get_latest_checkpoint_sequence_number()
+                        .await
+                        .expect("get_latest_checkpoint_sequence_number should not fail"),
+                }
+            }))
+            .await;
+
+        // The latest `latest_checkpoint` among all rpc servers
+        let max_checkpoint = end_checkpoints
+            .iter()
+            .max()
+            .expect("get_latest_checkpoint_sequence_number should not return empty");
+
+        debug!("GetCheckpoints({}, {:?})", op.start, max_checkpoint);
+
+        // TODO(chris): read `cross_validate` from config
+        let cross_validate = true;
+
+        for seq in op.start..=*max_checkpoint {
+            let transaction_digests: Arc<Mutex<DashSet<TransactionDigest>>> =
+                Arc::new(Mutex::new(DashSet::new()));
+            let checkpoints = join_all(clients.iter().enumerate().map(|(i, client)| {
+                let transaction_digests = transaction_digests.clone();
+                let end_checkpoint_for_clients = end_checkpoints.clone();
+                async move {
+                    if end_checkpoint_for_clients[i] < seq {
+                        // TODO(chris) log actual url
+                        warn!(
+                            "The RPC server corresponding to the {i}th url has a outdated checkpoint number {}.\
+                            The latest checkpoint number is {seq}",
+                            end_checkpoint_for_clients[i]
+                        );
+                        return None;
+                    }
+
+                    match client
+                        .read_api()
+                        .get_checkpoint(CheckpointId::SequenceNumber(<BigInt>::from(seq)))
+                        .await {
+                        Ok(t) => {
+                            if t.sequence_number != <BigInt>::from(seq) {
+                                error!("The RPC server corresponding to the {i}th url has unexpected checkpoint sequence number {}, expected {seq}", t.sequence_number,);
+                            }
+                            for digest in t.transactions.iter() {
+                                transaction_digests.lock().await.insert(*digest);
+                            }
+                            Some(t)
+                        },
+                        Err(err) => {
+                            error!("Failed to fetch checkpoint {seq} on the {i}th url: {err}");
+                            None
+                        }
+                    }
+                }
+            }))
+                .await;
+
+            let transaction_digests = transaction_digests
```

Comment:
Good question! 

First, I should have used `Arc<DashMap>` instead of `Arc<Mutex<DashMaps>>` according to the [documentation](https://github.com/xacrimon/dashmap).

In this case, it's not necessary to use DashSet, but the performance difference between DashSet and HashSet should be small, and later on we might need to use dashset for aggregating all addresses across threads